Installation issue:  ioapi

### Steps to reproduce

```console
spack install ioapi
```

### Error message

```console
lib/spack/spack/installer.py:1699 ==> [2022-01-05-10:36:24.399625] ioapi: Building ioapi-3.2-hdcvezwvxjjyo42ue5zfsmzhi2tfo46h [MakefilePackage]
lib/spack/spack/installer.py:1767 ==> [2022-01-05-10:36:24.406005] ioapi: Executing phase: 'edit'
lib/spack/spack/installer.py:1767 ==> [2022-01-05-10:36:24.459961] ioapi: Executing phase: 'build'
lib/spack/spack/error.py:55 ==> [2022-01-05-10:36:24.481186] Error: ProcessError: Command exited with status 2:
    'make' '-j2'

1 error found in build log:
     3    lib/spack/llnl/util/filesystem.py:163 ==> [2022-01-05-10:36:24.449945] FILTER FILE: Makefile [replacing "^INS
          TALL.*"]
     4    lib/spack/llnl/util/filesystem.py:163 ==> [2022-01-05-10:36:24.450980] FILTER FILE: Makefile [replacing "^BIN
          INST.*"]
     5    lib/spack/llnl/util/filesystem.py:163 ==> [2022-01-05-10:36:24.451977] FILTER FILE: Makefile [replacing "^LIB
          INST.*"]
     6    lib/spack/spack/installer.py:1767 ==> [2022-01-05-10:36:24.459961] ioapi: Executing phase: 'build'
     7    lib/spack/spack/util/executable.py:188 ==> [2022-01-05-10:36:24.461607] 'make' '-j2'
     8    (cd /tmp/vmware/spack-stage/spack-stage-ioapi-3.2-hdcvezwvxjjyo42ue5zfsmzhi2tfo46h/spack-src/ioapi   ; make f
          ixed_src)
  >> 9    make: *** No rule to make target '/tmp/vmware/spack-stage/spack-stage-ioapi-3.2-hdcvezwvxjjyo42ue5zfsmzhi2tfo
          46h/spack-src/ioapi/Makefile', needed by 'configure'.  Stop.
     10   make: *** Waiting for unfinished jobs....
     11   make[1]: warning: jobserver unavailable: using -j1.  Add '+' to parent make rule.
     12   make[1]: Entering directory '/tmp/vmware/spack-stage/spack-stage-ioapi-3.2-hdcvezwvxjjyo42ue5zfsmzhi2tfo46h/s
          pack-src/ioapi'
     13   make[1]: Nothing to be done for 'fixed_src'.
     14   make[1]: Leaving directory '/tmp/vmware/spack-stage/spack-stage-ioapi-3.2-hdcvezwvxjjyo42ue5zfsmzhi2tfo46h/sp
          ack-src/ioapi'

See build log for details:
  /tmp/vmware/spack-stage/spack-stage-ioapi-3.2-hdcvezwvxjjyo42ue5zfsmzhi2tfo46h/spack-build-out.txt

Traceback (most recent call last):
  File "/home/vmware/spack/lib/spack/spack/build_environment.py", line 914, in _setup_pkg_and_run
    return_value = function(pkg, kwargs)
  File "/home/vmware/spack/lib/spack/spack/installer.py", line 1776, in build_process
    phase(pkg.spec, pkg.prefix)
  File "/home/vmware/spack/lib/spack/spack/package.py", line 111, in phase_wrapper
    phase(spec, prefix)
  File "/home/vmware/spack/lib/spack/spack/build_systems/makefile.py", line 82, in build
    inspect.getmodule(self).make(*self.build_targets)
  File "/home/vmware/spack/lib/spack/spack/build_environment.py", line 138, in __call__
    return super(MakeExecutable, self).__call__(*args, **kwargs)
  File "/home/vmware/spack/lib/spack/spack/util/executable.py", line 223, in __call__
    raise ProcessError('Command exited with status %d:' %
spack.util.executable.ProcessError: Command exited with status 2:
    'make' '-j2'
lib/spack/spack/installer.py:1331 ==> [2022-01-05-10:36:24.492536] Flagging ioapi-3.2-hdcvezwvxjjyo42ue5zfsmzhi2tfo46h as failed: ProcessError: Command exited with status 2:
    'make' '-j2'
lib/spack/spack/main.py:777 ==> [2022-01-05-10:36:24.498000] ChildError: ProcessError: Command exited with status 2:
    'make' '-j2'
```

### Information on your system

* **Spack:** 0.16.1-2558-ecb7d6dca1
* **Python:** 3.8.10
* **Platform:** linux-ubuntu20.04-skylake_avx512
* **Concretizer:** original

### General information

- [X] I have run `spack debug report` and reported the version of Spack/Python/Platform
- [X] I have searched the issues of this repo and believe this is not a duplicate
- [X] I have run the failing commands in debug mode and reported the output

@caldodge I like @IncubatorShokuhou's idea of using the IOAPI git repository.  For making adjustments to the build and linking process, it actually may make more sense to add edge cases to the git revisions of IOAPI's master branch instead of relying on the 3.2 release tag.  Perhaps spack has versioning logic of knowing what git reversion in a branch is more recent than another.

Fault reproduced on an Intel Westmere machine running and Ubuntu 18 with the default gcc 7.5.0 :

```console
$ date --utc
Tue May 24 19:22:52 UTC 2022
$ cd ~/spack
$ git pull
...
$ spack debug report                                                                        
* **Spack:** 0.18.0.dev0 (f2a81af70e1d70828a579ea325a0699d89c7a420)
* **Python:** 3.6.9                                                                                            
* **Platform:** linux-ubuntu18.04-westmere                                                                     
* **Concretizer:** clingo
$ source ~/spack/share/spack/setup-env.sh
$ spack spec -I ioapi                                                                       
Input spec                                                                                                     
--------------------------------                                                                               
 -   ioapi                                                                                                     
                                                                                                               
Concretized                                                                                                    
--------------------------------                                                                               
 -   ioapi@3.2%gcc@7.5.0 arch=linux-ubuntu18.04-westmere                                                
[+]      ^netcdf-c@4.8.1%gcc@7.5.0~dap~fsync~hdf4~jna+mpi~parallel-netcdf+pic+shared patches=de556da arch=linux-ubuntu18.04-westmere
[+]          ^hdf5@1.12.2%gcc@7.5.0~cxx~fortran+hl~ipo~java+mpi+shared~szip~threadsafe+tools api=default build_type=RelWithDebInfo arch=linux-ubuntu18.04-westmere                                                            
[+]              ^cmake@3.23.1%gcc@7.5.0~doc+ncurses+ownlibs~qt build_type=Release arch=linux-ubuntu18.04-westmere                                                                                                            
[+]                  ^ncurses@6.2%gcc@7.5.0~symlinks+termlib abi=none arch=linux-ubuntu18.04-westmere
[+]                  ^openssl@1.1.1n%gcc@7.5.0~docs~shared certs=system arch=linux-ubuntu18.04-westmere
[+]                      ^zlib@1.2.12%gcc@7.5.0+optimize+pic+shared patches=0d38234 arch=linux-ubuntu18.04-westmere                                                                                                           
[+]              ^openmpi@4.1.3%gcc@7.5.0~atomics~cuda~cxx~cxx_exceptions~gpfs~internal-hwloc~java~legacylaunchers~lustre~memchecker~pmi+pmix+romio+rsh~singularity+static+vt+wrapper-rpath fabrics=none schedulers=none arch=linux-ubuntu18.04-westmere                                                                                     
[+]                  ^hwloc@2.7.1%gcc@7.5.0~cairo~cuda~gl~libudev+libxml2~netloc~nvml~opencl+pci~rocm+shared arch=linux-ubuntu18.04-westmere
[+]                      ^libpciaccess@0.16%gcc@7.5.0 arch=linux-ubuntu18.04-westmere
[+]                      ^libxml2@2.9.13%gcc@7.5.0~python arch=linux-ubuntu18.04-westmere
[+]                          ^libiconv@1.16%gcc@7.5.0 libs=shared,static arch=linux-ubuntu18.04-westmere
[+]                          ^xz@5.2.5%gcc@7.5.0~pic libs=shared,static arch=linux-ubuntu18.04-westmere
[+]                  ^libevent@2.1.12%gcc@7.5.0+openssl arch=linux-ubuntu18.04-westmere
[+]                  ^numactl@2.0.14%gcc@7.5.0 patches=4e1d78c,62fc8a8,ff37630 arch=linux-ubuntu18.04-westmere
[+]                      ^autoconf@2.69%gcc@7.5.0 patches=35c4492,7793209,a49dd5b arch=linux-ubuntu18.04-westmere
[+]                          ^m4@1.4.19%gcc@7.5.0+sigsegv patches=9dc5fbd,bfdffa7 arch=linux-ubuntu18.04-westmere
[+]                              ^libsigsegv@2.13%gcc@7.5.0 arch=linux-ubuntu18.04-westmere
[+]                          ^perl@5.34.1%gcc@7.5.0+cpanm+shared+threads arch=linux-ubuntu18.04-westmere
[+]                              ^berkeley-db@18.1.40%gcc@7.5.0+cxx~docs+stl patches=b231fcc arch=linux-ubuntu18.04-westmere
[+]                              ^bzip2@1.0.8%gcc@7.5.0~debug~pic+shared arch=linux-ubuntu18.04-westmere
[+]                              ^gdbm@1.19%gcc@7.5.0 arch=linux-ubuntu18.04-westmere
[+]                                  ^readline@8.1%gcc@7.5.0 arch=linux-ubuntu18.04-westmere
[+]                      ^automake@1.16.5%gcc@7.5.0 arch=linux-ubuntu18.04-westmere
[+]                      ^libtool@2.4.7%gcc@7.5.0 arch=linux-ubuntu18.04-westmere
[+]                  ^openssh@9.0p1%gcc@7.5.0 arch=linux-ubuntu18.04-westmere
[+]                      ^libedit@3.1-20210216%gcc@7.5.0 arch=linux-ubuntu18.04-westmere
[+]                  ^pkgconf@1.8.0%gcc@7.5.0 arch=linux-ubuntu18.04-westmere
[+]                  ^pmix@4.1.2%gcc@7.5.0~docs+pmi_backwards_compatibility~restful arch=linux-ubuntu18.04-westmere
[+]      ^netcdf-fortran@4.5.4%gcc@7.5.0~doc+pic+shared arch=linux-ubuntu18.04-westmere
[+]      ^sed@4.8%gcc@7.5.0 arch=linux-ubuntu18.04-westmere
$ spack install ioapi
...
==> Installing ioapi-3.2-kgxbdt6obwmphpmzxyx5fwn5mlezlyvn
==> No binary for ioapi-3.2-kgxbdt6obwmphpmzxyx5fwn5mlezlyvn found: installing from source
==> Using cached archive: /home/pnanda/spack/var/spack/cache/_source-cache/archive/0a/0a3cbf236ffbd9fb5f6509e35
308c3353f1f53096efe0c51b84883d2da86924b.tar.gz
==> No patches needed for ioapi
==> ioapi: Executing phase: 'edit'
==> ioapi: Executing phase: 'build'
==> Error: ProcessError: Command exited with status 2:
    'make' '-j12'

1 error found in build log: 
     3    ==> [2022-05-24-16:09:23.439880] FILTER FILE: Makefile [replacing "^INSTALL.*"]
     4    ==> [2022-05-24-16:09:23.441537] FILTER FILE: Makefile [replacing "^BININST.*"]
     5    ==> [2022-05-24-16:09:23.443089] FILTER FILE: Makefile [replacing "^LIBINST.*"]
     6    ==> ioapi: Executing phase: 'build'
     7    ==> [2022-05-24-16:09:23.475484] 'make' '-j12'
     8    (cd /tmp/pnanda/spack-stage/spack-stage-ioapi-3.2-kgxbdt6obwmphpmzxyx5fwn5mlezlyvn/spack-src/ioapi  
           ; make fixed_src)
  >> 9    make: *** No rule to make target '/tmp/pnanda/spack-stage/spack-stage-ioapi-3.2-kgxbdt6obwmphpmzxyx5
          fwn5mlezlyvn/spack-src/ioapi/Makefile', needed by 'configure'.  Stop.
     10   make: *** Waiting for unfinished jobs....
     11   make[1]: warning: jobserver unavailable: using -j1.  Add '+' to parent make rule.
     12   make[1]: Entering directory '/mnt/aux/tmp/pnanda/spack-stage/spack-stage-ioapi-3.2-kgxbdt6obwmphpmzx
          yx5fwn5mlezlyvn/spack-src/ioapi'
     13   make[1]: Nothing to be done for 'fixed_src'.
     14   make[1]: Leaving directory '/mnt/aux/tmp/pnanda/spack-stage/spack-stage-ioapi-3.2-kgxbdt6obwmphpmzxy
          x5fwn5mlezlyvn/spack-src/ioapi'

See build log for details:
  /tmp/pnanda/spack-stage/spack-stage-ioapi-3.2-kgxbdt6obwmphpmzxyx5fwn5mlezlyvn/spack-build-out.txt
```

* [spack-build-out.txt](https://github.com/spack/spack/files/8765916/spack-build-out.txt)
* [spack-build-env.txt](https://github.com/spack/spack/files/8765917/spack-build-env.txt)
 
Trying with the current IOAPI git version.

```console
$ cd
$ mkdir -p tmp/
$ cd ~/tmp/
$ git clone https://github.com/cjcoats/ioapi-3.2
...
$ cd ioapi-3.2/
$ spack dev-build ioapi@3.2
...
$ git diff
diff --git a/Makefile.template b/Makefile.template
index 599c868..23ef85c 100644
--- a/Makefile.template
+++ b/Makefile.template
@@ -189,7 +189,7 @@ SEDCMD = \
 #      ----------   I/O API Build System directory definitions  --------
 
 VERSION    = 3.2-${CPLMODE}
-BASEDIR    = ${PWD}
+BASEDIR = /home/pnanda/tmp/ioapi-3.2
 NCFLIBS    = -lnetcdff -lnetcdf
 IODIR      = $(BASEDIR)/ioapi
 FIXDIR     = $(IODIR)/fixed_src
$ 
```

So the issue is only `BASEDIR` is patched but not `INSTALL`, `BININST` or `LIBINST` because of revision cjcoats/ioapi-3.2@55574be changes to `Makefile.template` on March 25, 2020:

```console
$ git log --patch Makefile.template
$ GIT_PAGER=cat git log -1 --patch 55574be Makefile.template
```

This seems simple enough to fix against the current tarball.  I'll need to discuss on the spack mailing list or slack about versioning against git to considering how what they think of that idea.

--- a/repos/spack_repo/builtin/packages/ioapi/package.py
+++ b/repos/spack_repo/builtin/packages/ioapi/package.py
@@ -13,20 +13,79 @@ class Ioapi(MakefilePackage):
     """Models-3/EDSS Input/Output Applications Programming Interface."""
 
     homepage = "https://www.cmascenter.org/ioapi/"
-    url = "https://www.cmascenter.org/ioapi/download/ioapi-3.2.tar.gz"
-    maintainers("omsai")
-    # This checksum is somewhat meaningless because upstream updates the tarball
-    # without incrementing the version despite requests no to do this.
-    # Therefore the checksum fails everytime upstream silently updates the
-    # source tarball (#28247).  This also means that one must test for breaking
-    # changes when updating the checksum and avoid #22633.
-    version("3.2", sha256="0a3cbf236ffbd9fb5f6509e35308c3353f1f53096efe0c51b84883d2da86924b")
+    git = "https://github.com/cjcoats/ioapi-3.2.git"
 
-    depends_on("c", type="build")  # generated
-    depends_on("fortran", type="build")  # generated
-    depends_on("netcdf-c@4:")
-    depends_on("netcdf-fortran@4:")
+    maintainers("omsai")
+
+    license("GPL-2.0-or-later", checked_by="omsai")
+
+    # The two -develop versions below without date suffixes are "rolling
+    # releases":
+    # https://www.cmascenter.org/ioapi/documentation/all_versions/html/AVAIL.html#build
+    # If you installed ioapi@3.2 previously it was a rolling release that
+    # caused checksums to fail (#22633, #28247).  Until recently, upstream did
+    # not freeze tarballs and now freezes them with a date suffix.
+    version("4.0-develop", branch="ioapi-4.0")
+    version("3.2-develop", branch="master")
+    # Dated suffix versions are now pulled from GitHub.  The dates below are
+    # based on the, the GitHub tag release date, because there is no other
+    # reliable version:
+    # 1. Most of upstream's tagged versions are missing a leading zero and are
+    #    therefore not in ascending numerical order which break versioning
+    #    relationships.
+    # 2. The tarball VERSION.txt is not reliably updated.
+    version(
+        "3.2.20200828",
+        commit="ef5d5f4e112c249b593b19426421f25d79ae094b",
+        preferred=True,
+    )
+    version(
+        "3.2.20200714",
+        commit="6ebb47e96db3b641af63ee5f853c943b596a1268",
+    )
+    version(
+        "3.2.20200420",
+        commit="4017280cc656993a5be50f0da9287a56166da22b",
+    )
+
+    # MPI support is not yet well supported in this spack package and
+    # may fail to build.
+    variant("mpi", default=False, description="Enable MPI support")
+
+    depends_on("c", type="build")
+    depends_on("fortran", type="build")
     depends_on("sed", type="build")
+
+    depends_on("mpi", when="+mpi")
+    depends_on("netcdf-c@4.4.2: ~mpi +parallel-netcdf", when="+mpi")
+    depends_on("netcdf-c@4.4.2: ~mpi", when="~mpi")
+    depends_on("netcdf-fortran@4.4.2:")
+
+    # Supporting compilers by mapping to their Makeinclude files
+    # requires additional work and testing; the package naming scheme
+    # is inconsistent perhaps because it's meant more for human
+    # editing.
+    conflicts(
+        "^intel-oneapi-compilers",
+        msg="Update this spack package to support oneapi",
+    )
+    conflicts("%nvhpc", msg="Update this spack package to support pgi")
+    # There is no evidence of support for LLVM even on macOS.
+    conflicts("%llvm", msg="IOAPI does not support LLVM")
+
+    # Parallel build will fail with this error:
+    #
+    # Fatal Error: Cannot open module file 'm3utilio.mod' for reading at (1):
+    # No such file or directory
+    parallel = False
+
+    def setup_build_environment(self, env):
+        # The BIN environmental variables needs to be set in addition
+        # to being written into the top-level Makefile.
+        BIN = "Linux2_x86_64gfort"
+        if self.spec.satisfies("+mpi"):
+            BIN += "mpi"
+        env.set("BIN", BIN)
 
     def edit(self, spec, prefix):
         # No default Makefile bundled; edit the template.
@@ -41,6 +100,9 @@ CPLMODE = nocpl
 \\1
         """.strip(),
         )
+        BIN = "Linux2_x86_64gfort"
+        if self.spec.satisfies("+mpi"):
+            BIN += "mpi"
         makefile.filter(
             "^BASEDIR.*",
             (
@@ -56,15 +118,25 @@ BININST = """
                 + """
 LIBINST = """
                 + prefix.lib
-                + """
-BIN = Linux2_x86_64
+                + f"""
+BIN = {BIN}
         """
             ).strip(),
         )
         # Fix circular dependency bug for generating subdirectory Makefiles.
         makefile.filter("^configure:.*", "configure:")
+        # Fix hard-coded fortran mpi compiler.
+        if self.spec.satisfies("+mpi"):
+            makeinclude = FileFilter(f"ioapi/Makeinclude.{BIN}")
+            makeinclude.filter("mpicc", f"{self.spec['mpi'].mpicc}")
+            makeinclude.filter("mpif90", f"{self.spec['mpi'].mpifc}")
         # Generate the subdirectory Makefiles.
         make("configure")
+
+    def flag_handler(self, name: str, flags: List[str]):
+        if name == "fflags" and self.spec.satisfies("%fortran=gcc@10:"):
+            flags.append("-fallow-argument-mismatch")
+        return (flags, None, None)
 
     def install(self, spec, prefix):
         make("install")

--- a/repos/spack_repo/builtin/packages/ioapi/package.py
+++ b/repos/spack_repo/builtin/packages/ioapi/package.py
@@ -34,19 +34,9 @@ class Ioapi(MakefilePackage):
     #    therefore not in ascending numerical order which break versioning
     #    relationships.
     # 2. The tarball VERSION.txt is not reliably updated.
-    version(
-        "3.2.20200828",
-        commit="ef5d5f4e112c249b593b19426421f25d79ae094b",
-        preferred=True,
-    )
-    version(
-        "3.2.20200714",
-        commit="6ebb47e96db3b641af63ee5f853c943b596a1268",
-    )
-    version(
-        "3.2.20200420",
-        commit="4017280cc656993a5be50f0da9287a56166da22b",
-    )
+    version("3.2.20200828", commit="ef5d5f4e112c249b593b19426421f25d79ae094b", preferred=True)
+    version("3.2.20200714", commit="6ebb47e96db3b641af63ee5f853c943b596a1268")
+    version("3.2.20200420", commit="4017280cc656993a5be50f0da9287a56166da22b")
 
     # MPI support is not yet well supported in this spack package and
     # may fail to build.
@@ -65,10 +55,7 @@ class Ioapi(MakefilePackage):
     # requires additional work and testing; the package naming scheme
     # is inconsistent perhaps because it's meant more for human
     # editing.
-    conflicts(
-        "^intel-oneapi-compilers",
-        msg="Update this spack package to support oneapi",
-    )
+    conflicts("^intel-oneapi-compilers", msg="Update this spack package to support oneapi")
     conflicts("%nvhpc", msg="Update this spack package to support pgi")
     # There is no evidence of support for LLVM even on macOS.
     conflicts("%llvm", msg="IOAPI does not support LLVM")

